### ISSUE TITLE
Infrastructure Monitoring: Migrate DataDog to GovCloud instance

### ISSUE THREAD

**Issue by jhouse-solvd** [department-of-veterans-affairs/va.gov-team#23513]
**Note: This initiative should move to recently complete. @jhouse-solvd 9/14**

## Description 
DataDog is moving to a dedicated installation in VA's GovCloud. The DOTS team has offered to help migrate dashboards, alerts, and monitors, but cannot migrate historical data. 

## Details

- The DOTS team is migrating DataDog from its current SaaS instance (datadoghq.com) to GovCloud (ddog-gov.com)
- Authentication methods for this service will change (will be using Ablevets Okta)
- Authentication to Okta will be with a username and password
- DataDog engineers are available in the VA OIT DevOps Slack workspace to discuss technical details 
- **All historic data will be lost and unable to be migrated** 
- DataDog agents are being re-deployed to point to the new instance
- There will be a new subdomain: vagov.ddog-gov.com
- Once the account is migrated, the existing account will remain active for a short time but will be sunset by June 30, 2022
- The Platform Infrastructure Team is meeting w/ the DOTS team to gather add'l details 

## Add'l considerations
- Authentication method will apply to the whole DataDog account, not per user.
- DataDog in GovCloud will still be accessible from anywhere (even off of the VA network

## Definition of done

- Platform Crew users are able to access the new instance
- DataDog agents are sending metrics to the new instance
- DataDog monitors are sending notifications 
  - to Slack
  - to PagerDuty 
- DataDog resources have been migrated to the new instance 
- Documentation exists that highlights the following: 
  - How to request access to the new DataDog instance 
  - How to access the new DataDog instance
  - How to migrate resources to the new DataDog instance

**Comment by jhouse-solvd** [comment]
The Operations team is not currently clear on the amount of work that will be required, as much of the migration will be handled by the DOTS team directly. Once we have additional information - likely once the migration begins - we will update this ticket accordingly (including communicating any plans and potential impact).

**Comment by jhouse-solvd** [comment]
Needs to be refined in preparation for Q2 planning. 

The initial phase of work will be non-technical, and focused on communication w/ DOTS to have a plan and timeline in place.

**Comment by jhouse-solvd** [comment]
Mike Chelen and I are communicating with DOTS via va.gov email to iron out details. 

Once we have those sorted, we should be ready to plan work for an upcoming sprint.

**Comment by jhouse-solvd** [comment]
## Open questions 

### Administrative

- Have we decided on an org / sub-org structure?
- Can we extend the grace period until the end of June? There’s been feedback that having overlap would be best
- Who has DOTS invited to OKTA thus far? And how did it determine the list of invitees? _It was mentioned during our sync w/ DOTS on 3/31 that the platform would provide a list of a) admins and b) read-only users, but some folks have already received OKTA invites_

### Technical

- Can we send metrics to two DD instances at once? This will inform the implementation plan
- When can we expect to be assigned a DD engineer? _We’re hoping to do bulk export/import of dashboards. This, too, will inform the implementation plan_

**Comment by mchelen-gov** [comment]
> ## Open questions
> ### Administrative
> * Have we decided on an org / sub-org structure?

Not yet, we are waiting for guidance from DOTS and DD team on the options & tradeoffs.

> * Can we extend the grace period until the end of June? There’s been feedback that having overlap would be best

Agree that period of overlap is necessary. We can't set a firm cutover date until we know when the new instance will be working.

> * Who has DOTS invited to OKTA thus far? And how did it determine the list of invitees? _It was mentioned during our sync w/ DOTS on 3/31 that the platform would provide a list of a) admins and b) read-only users, but some folks have already received OKTA invites_

I'm not sure.

> 
> ### Technical
> * Can we send metrics to two DD instances at once? This will inform the implementation plan

Good question, I am not sure.

> * When can we expect to be assigned a DD engineer? _We’re hoping to do bulk export/import of dashboards. This, too, will inform the implementation plan_

I don't know, this is a good question for DOTS/DD.

**Comment by jhouse-solvd** [comment]
**Meeting planned w/ DOTS for 5/19**. @ph-One and @jhouse-solvd plan to be in attendance. 

_In addition to the questions raised above, @mleclerc00 mentioned the following:_ 

- getting metrics pointed to the new instance should be somewhat simple, ie provide a new key and URL to the agents
- a greater effort is likely to move private locations (and all of the synthetics using them)
- most of the monitors are in terraform (this is good!) 
- dashboards are not in Terraform (yet; b/c of effort associated w/ formatting)
- integrations currently in use for messaging/alerting: ie slack and pagerduty

These are great points of clarification that we intend to address during tomorrow's call. :)

**Comment by jhouse-solvd** [comment]
Met w/ DOTS & DataDog personnel 5/19 @ 1500 ET along w/ @ph-One & @mleclerc00 

We'll synthesize notes and link to this issue today or tomorrow.

**Comment by jhouse-solvd** [comment]
**Notes from call w/ DOTS & DataDog personnel on May 19, 2022:**

[2022-05-19: Meeting w/ DOTS and DataDog personnel](https://vfs.atlassian.net/wiki/spaces/OT/pages/2204041225/2022-05-19+Meeting+w+DOTS+and+DataDog+personnel)

**TL;DR**

- The Platform needs to fully **transition to the new GovCloud account no later than June 30,** 2022 
- It is recommended that there be **one organization for Platform and VFS engineers**
- The Platform Infrastructure team will **configure DataDog agents to send metrics to the new and old accounts simultaneously** to be confident that things are configured properly in the new account
- **Most things can be moved in bulk, however, others must be recreated** in the new account
- The Platform Infrastructure team can **discuss details with DOTS and DataDog personnel in the VA OIT DevOps Slack** workspace

**Comment by jhouse-solvd** [comment]
Link to DOTS Jira to request new DataDog users: https://vajira.max.gov/servicedesk/customer/portal/1/create/18

**Comment by jhouse-solvd** [comment]
@mchelen-gov 

- I'm not sure about the exact specifics, but we can create custom roles that give our org the ability to create new roles with unique permission sets
- re: Billing, I would recommend raising that question in #public-datadog in VA OIT DevOps Slack

**Comment by ianhundere** [comment]
**process for importing / syncing datadog resources**

using `datadog-sync` the following resources can be synced across datadog instances.
```
roles
users
synthetics_private_locations
synthetics_tests
synthetics_global_variables
monitors
downtimes
service_level_objectives
slo_corrections
dashboards
dashboard_lists
logs_custom_pipelines
notebooks
host_tags
logs_metrics
metric_tag_configurations
```

and [these resources often rely on one another](https://github.com/DataDog/datadog-sync-cli#best-practices).

**using `datadog-sync`**

the two main actions of `datadog-sync` are `import` (saves datadog resources locally) and `sync` (send local cache of dd resources to specified destination). while you can use the `--config` flag to pass a config file, the below demonstrates running the tool with all necessary information via the cli for clarity.

the following are necessary for running an import:

`datadog-sync import --source-api-key <source-api-key> --source-app-key <source-app-key>`

the following are necessary for running a sync:

`datadog-sync sync --destination-api-url https://vagov.ddog-gov.com --destination-api-key <source-api-key> --destination-app-key <source-app-key>  --source-api-key <source-api-key> --source-app-key <source-app-key>`

the reason the destination url is necessary, but not the source url is that each are `https://api.datadoghq.com` by default.

**importing**

the tool provides a `--filter` flag, but in my experience i also found i needed to use the `--resources` flag as well otherwise i would `import` all resources. i also had to audit what resources i would be importing. those resources ended up being the following:

`synthetics_tests`
`synthetics_global_variables`
`monitors`
`dashboards`

thankfully, many of these resources had tags which made importing a little easier. the following commands were ran to import each of these:
- `datadog-sync import --source-api-key <source-api-key> --source-app-key <source-app-key> --resources synthetics_tests --filter="Type=synthetics_tests;Name=tags;Value=team:identity"`
- `datadog-sync import --source-api-key <source-api-key> --source-app-key <source-app-key> --resources synthetics_global_variables --filter="Type=synthetics_global_variables;Name=tags;Value=team:identity`
- `datadog-sync import --source-api-key <source-api-key> --source-app-key <source-app-key> --resources monitors --filter="Type=Monitors;Name=tags;Value=team:identity"`
- `datadog-sync import --source-api-key <source-api-key> --source-app-key <source-app-key> --resources dashboards --filter="Type=dashboards;Name=title;Value=VSP Identity Monitor Dashboard`

the only resource the was imported that does not support tags is the dashboard, so i queried it based off its name. remember, if you had special characters such as `[` `]` you will need to escape them like so `\[EKS\]`.

once import the following file structure will be created:
```
resources
├── destination
│   ├── dashboard_lists.json
│   ├── dashboards.json
│   ├── downtimes.json
│   ├── host_tags.json
│   ├── logs_custom_pipelines.json
│   ├── logs_metrics.json
│   ├── metric_tag_configurations.json
│   ├── monitors.json
│   ├── notebooks.json
│   ├── roles.json
│   ├── service_level_objectives.json
│   ├── slo_corrections.json
│   ├── synthetics_global_variables.json
│   ├── synthetics_private_locations.json
│   ├── synthetics_tests.json
│   └── users.json
└── source
    ├── dashboards.json
    ├── monitors.json
    ├── synthetics_global_variables.json
    └── synthetics_tests.json
```

notice that under `destination` all resource files are created, but are empoty (e.g. `{}`).

**syncing**

syncing resource requires much more effort and thought as i will explain below.

_private locations_

again, it is important to note that each resource can be reliant on another. `synthetics_tests` rely on `synthetics_private_locations`, but since the datadog instance these resources will be moved to have a different set of not only "Private Locations", but also "Default Locations." the new gov instance only includes one aws region, and that is `us-gov-west-1`. so not only will the `synthetics_private_locations` need to changed on the `synthetics_tests`, but also any default locations as well (e.g. `us-east-1` to `us-gov-west-1`). 

_global variables for synthetic tests_

these will need to be removed from synthetic tests because when you import them they’ll not have the appropriate value since it's a secure value (e.g. `"secure": true`). if you do include them, you'll receive an error like `{"errors": ["Validation failed for input."]}`. so to reiterate:

- remove any `synthetics_global_variables` that are secure from `synthetics_tests` before syncing
- once the `synthetics_tests`  have been synced, add the values for the `synthetics_global_variables`
- then add the `synthetics_global_variables` back to the synthetics_tests

**Comment by jhouse-solvd** [comment]
Ian's notes (above) have been copied to Confluence and linked to #42961

_See here:_ [Migrate DataDog resources to a new organization](https://vfs.atlassian.net/wiki/spaces/OT/pages/2235859230/Migrate+DataDog+resources+to+new+instance)

**Comment by jhouse-solvd** [comment]
We're tracking feedback and technical issues that Platform Crew teams encouter during migration in #43094

**Comment by mchelen-gov** [comment]
Once Platform teams have migrated to Gov instance, all non-Infrastructure users should be removed from Admin role.

**Comment by ph-One** [comment]
> Once Platform teams have migrated to Gov instance, all non-Infrastructure users should be removed from Admin role.

@mchelen-gov  The current admin list is mostly TL's, PM's, or DevOps, with a couple that I'm unsure of. Would you like me to demote all non-infra (or DevOps) to the `standard` role, or just have me follow-up with the non-TL/PM/DevOps on the list to see if they should be removed? I believe this would be the final bit of work necessary to close out this epic

**Comment by oseasmoran73** [comment]
Monitoring and alerting has been migrated from previous into current datadog and actively being used. See [example](https://dsva.slack.com/archives/C30LCU8S3/p1663000518985529)

Per DoD, all have been met

**Comment by jhouse-solvd** [comment]
A few important notes: 
- Monitoring with Datadog will be offered to VFS teams in #33081. 
- _Before offering Datadog to VFS teams_, custom permissions should be developed using Datadog's role-based access control (see #47066)
- No VFS user-facing documentation has been added to the Platform Website as part of this migration  
- Operational & Platform Crew user-facing documentation exists in Confluence [here](https://vfs.atlassian.net/wiki/spaces/OT/pages/2236612646/Monitor+things+with+DataDog)

**Comment by jhouse-solvd** [comment]
The Definition of Done has been met. Documentation and follow-up issues are linked in the comments above.

This issue can be closed. 🚀
